xmake/rules/c++/modules/xmake.lua

```diff
@@ -21,45 +21,122 @@
 -- define rule: c++.build.modules
 rule("c++.build.modules")
     set_extensions(".mpp", ".mxx", ".cppm", ".ixx")
+    add_deps("c++.build.modules.dependencies")
+    add_deps("c++.build.modules.builder")
+    add_deps("c++.build.modules.install")
+
     on_config(function (target)
         -- we disable to build across targets in parallel, because the source files may depend on other target modules
         -- @see https://github.com/xmake-io/xmake/issues/1858
-        local target_with_modules
-        for _, dep in ipairs(target:orderdeps()) do
-            local sourcebatches = dep:sourcebatches()
-            if sourcebatches and sourcebatches["c++.build.modules"] then
-                target_with_modules = true
-                break
-            end
-        end
+        local common = import("modules_support.common")
+
+        local target_with_modules = target:sourcebatches()["c++.build.modules"] and common.contains_modules(target) or false
         if target_with_modules then
             -- @note this will cause cross-parallel builds to be disabled for all sub-dependent targets,
             -- even if some sub-targets do not contain C++ modules.
             --
             -- maybe we will have a more fine-grained configuration strategy to disable it in the future.
             target:set("policy", "build.across_targets_in_parallel", false)
```

Comment:
`modules_support.load_parent` only for `common.contains_modules_in_targetdeps` and `across_targets_in_parallel`, It is only to solve this issue

```lua

target("mod")
  set_kind("static")
  add_files("mod.mpp", "mod.cpp")

target("main")
  set_kind("binary")
  add_deps("mod")
  add_files("main.cpp")  -- no .mpp, but it depend on target with modules
```

so we need improve it.

```lua
        -- get modules support
        local modules_support = common.modules_support(target)

        -- check C++20 module support
        modules_support.check_module_support(target)

        -- mark this target with modules
        target:set("cxx.has_modules", true)

        -- we disable to build across targets in parallel, because the source files may depend on other target modules
        -- @see https://github.com/xmake-io/xmake/issues/1858
        if common.contains_modules_in_targetdeps(target) then
            -- @note this will cause cross-parallel builds to be disabled for all sub-dependent targets,
            -- even if some sub-targets do not contain C++ modules.
            --
            -- maybe we will have a more fine-grained configuration strategy to disable it in the future.
            target:set("policy", "build.across_targets_in_parallel", false)

            -- load parent
            modules_support.load_parent(target, opt)
        end
```

Reply:
oh, I was wrong about that. maybe it should be 

```lua

        -- we disable to build across targets in parallel, because the source files may depend on other target modules
        -- @see https://github.com/xmake-io/xmake/issues/1858
        if common.contains_modules(target) then
            -- @note this will cause cross-parallel builds to be disabled for all sub-dependent targets,
            -- even if some sub-targets do not contain C++ modules.
            --
            -- maybe we will have a more fine-grained configuration strategy to disable it in the future.
            target:set("policy", "build.across_targets_in_parallel", false)

            -- get modules support
            local modules_support = common.modules_support(target)

            -- check C++20 module support
            modules_support.check_module_support(target)

            -- mark this target with modules
            target:data_set("cxx.has_modules", true)

            -- load parent
            modules_support.load_parent(target, opt)
        end
```